includes/specials/SpecialConcepts.php

```diff
@@ -42,27 +43,40 @@ public function __construct() {
 	 *
 	 * @return DIWikiPage[]
 	 */
-	public function getResults( $limit, $from, $until ) {
-		$description = new NamespaceDescription( SMW_NS_CONCEPT );
-		$query = SMWPageLister::getQuery( $description, $limit, $from, $until );
-
-		$results = $this->getStore()->getQueryResult( $query )->getResults();
-
-		foreach ( $results as $key => $subject ) {
-
-			// Don't display subobjects like _QUERY... on the list
-			if ( $subject->getSubobjectName() !== '' ) {
-				unset( $results[$key] );
-			}
-
-			// Avoid concepts that are mentioned in a query but doesn't
-			// exist as real page
-			if ( !$subject->getTitle()->exists() ) {
-				unset( $results[$key] );
-			}
+	public function getResults( $limit, $offset, $from, $until ) {
+
+		$connection = $this->getStore()->getConnection( 'mw.db' );
+		$results = [];
+
+		$fields = [
+			'smw_id',
+			'smw_title'
+		];
+
+		$conditions = [
+			'smw_namespace' => SMW_NS_CONCEPT,
+			'smw_iw' => '',
+			'smw_subobject' => ''
+		];
+
+		$options = [
+			'LIMIT' => $limit + 1,
+			'OFFSET' => $offset,
+		];
+
+		$res = $connection->select(
+			$connection->tableName( SQLStore::ID_TABLE ),
+			$fields,
+			$conditions,
+			__METHOD__,
+			$options
+		);
+
+		foreach ( $res as $row ) {
+			$results[] = new DIWikiPage( $row->smw_title, SMW_NS_CONCEPT );
 		}
 
-		return array_values( $results );
+		return $results;
```

Comment:
I'm a big fan of minimizing scope and mutability. In this case `$results` is in scope for quite some time. You could move it down a bit. It can also be avoided by just doing a map rather than the imperative form. Even if maps in PHP are not very elegant, still better than the foreach loop IMO.

```php
return array_map(
    function( SuchType $row ) {
        return new DIWikiPage( $row->smw_title, SMW_NS_CONCEPT );
    },
    $res
);
```

$res and $row are also not super names. Surely they represent things in the SMW domain?

Reply:
Scope comment also applies to getting the `$connection`. I'd inline that and perhaps extract part of the query stuff into its own method.

Reply:
> $res and $row are also not super names. Surely they represent things in the SMW domain?

$res and $row are terms in connection with a DB result and while those names are debatable there are fairly common as references in MW [0].

> I'm a big fan of minimizing scope and mutability. In this case $results is in scope for quite some time.

We agree on that.

> return array_map(

This will not work because `$res` is an iterator which cannot be used directly with `array_map` [1, 2]. Of course, you could use `iterator_to_array` but then we resolve the iterator beforehand and that brings no benefit in terms of memory consumption compared to a `foreach` approach.

[0] https://github.com/wikimedia/mediawiki/search?utf8=%E2%9C%93&q=res&type=
[1] http://php.net/manual/en/function.array-map.php
[2] https://stackoverflow.com/questions/16273233/array-map-on-collection-with-array-interfaces

Reply:
Oh did not realize this thing was an interator. And of course PHP sucks balls when it comes to iterators :\

> $res and $row are terms in connection with a DB result and while those names are debatable there are fairly common as references in MW [0].

MW really is not the place to look for how to write clean code :D